test/client_mqtt5.js

```diff
@@ -13,29 +13,548 @@ describe('MQTT 5.0', function () {
+          messageId: 0,
+          topic: 'test2',
+          payload: 'Message',
+          qos: 0,
+          properties: { topicAlias: 1 }
+        })
+        // use topicAlias
+        serverClient.publish({
+          messageId: 0,
+          topic: '',
+          payload: 'Message',
+          qos: 0,
+          properties: { topicAlias: 1 }
+        })
+      })
+    }).listen(ports.PORTAND103)
+
+    client.on('message', function (topic, messagee, packet) {
+      switch (publishCount++) {
+        case 0:
+          assert.strictEqual(topic, 'test1')
+          assert.strictEqual(packet.topic, 'test1')
+          assert.strictEqual(packet.properties.topicAlias, 1)
+          break
+        case 1:
+          assert.strictEqual(topic, 'test1')
+          assert.strictEqual(packet.topic, '')
+          assert.strictEqual(packet.properties.topicAlias, 1)
+          break
+        case 2:
+          assert.strictEqual(topic, 'test2')
+          assert.strictEqual(packet.topic, 'test2')
+          assert.strictEqual(packet.properties.topicAlias, 1)
+          break
+        case 3:
+          assert.strictEqual(topic, 'test2')
+          assert.strictEqual(packet.topic, '')
+          assert.strictEqual(packet.properties.topicAlias, 1)
+          server103.close()
+          client.end(true, done)
+          break
+      }
+    })
+  })
+
+  it('registered topic alias should automatically used if autoUseTopicAlias is true', function (done) {
+    this.timeout(15000)
+
+    var opts = {
+      host: 'localhost',
+      port: ports.PORTAND103,
+      protocolVersion: 5,
+      autoUseTopicAlias: true
+    }
+    var client = mqtt.connect(opts)
+
+    var publishCount = 0
+    var server103 = new MqttServer(function (serverClient) {
+      serverClient.on('connect', function (packet) {
+        serverClient.connack({
+          reasonCode: 0,
+          properties: {
+            topicAliasMaximum: 3
+          }
+        })
+      })
+      serverClient.on('publish', function (packet) {
+        switch (publishCount++) {
+          case 0:
+            assert.strictEqual(packet.topic, 'test1')
+            assert.strictEqual(packet.properties.topicAlias, 1)
+            break
+          case 1:
+            assert.strictEqual(packet.topic, '')
+            assert.strictEqual(packet.properties.topicAlias, 1)
+            break
+          case 2:
+            assert.strictEqual(packet.topic, '')
+            assert.strictEqual(packet.properties.topicAlias, 1)
+            server103.close()
+            client.end(true, done)
+            break
+        }
+      })
+    }).listen(ports.PORTAND103)
+
+    client.on('connect', function () {
+      // register topicAlias
+      client.publish('test1', 'Message', { properties: { topicAlias: 1 } })
+      // use topicAlias
+      client.publish('', 'Message', { properties: { topicAlias: 1 } })
+      // use topicAlias by autoApplyTopicAlias
+      client.publish('test1', 'Message')
```

Comment:
so when you enable autoApplyTopicAlias it will substitute the alias for the message, and existing behaviour without it set to true is that on `client.publish('test1', 'Message')`, the topicAlias already registered wouldn't be used?

Reply:
Yes. If autoApplyTopicAlias is true, and the user wants to use Topic Alias, the user needs to set the empty TopicName and add Topic Alias property manually.